flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaRecordSerializationSchemaBuilder.java

```diff
@@ -0,0 +1,270 @@
+     */
+    public KafkaRecordSerializationSchema<IN> build() {
+        checkState(valueSerializationSchema != null, "No value serializer is configured.");
+        return new KafkaRecordSerializationSchema<>() {
+
+            @Override
+            public void open(
+                    SerializationSchema.InitializationContext context, KafkaSinkContext sinkContext)
+                    throws Exception {
+                valueSerializationSchema.open(context);
+                if (keySerializationSchema != null) {
+                    keySerializationSchema.open(context);
+                }
+                if (partitioner != null) {
+                    partitioner.open(
+                            sinkContext.getParallelInstanceId(),
+                            sinkContext.getNumberOfParallelInstances());
+                }
+            }
+
+            @Override
+            public ProducerRecord<byte[], byte[]> serialize(
+                    IN element, KafkaSinkContext context, Long timestamp) {
+                final String targetTopic = topic == null ? topicSelector.apply(element) : topic;
+                final byte[] value = valueSerializationSchema.serialize(element);
+                byte[] key = null;
+                if (keySerializationSchema != null) {
+                    key = keySerializationSchema.serialize(element);
+                }
+                final OptionalInt partition =
+                        partitioner != null
+                                ? OptionalInt.of(
+                                        partitioner.partition(
+                                                element,
+                                                key,
+                                                value,
+                                                targetTopic,
+                                                context.getPartitionsForTopic(targetTopic)))
+                                : OptionalInt.empty();
+                if (partition.isPresent()) {
+                    return new ProducerRecord<>(targetTopic, partition.getAsInt(), key, value);
+                }
+                if (key == null) {
+                    return new ProducerRecord<>(targetTopic, value);
+                }
+                return new ProducerRecord<>(targetTopic, key, value);
+            }
+        };
+    }
+
+    private static String exceedsSerializerConfigurations(String serializerType) {
+        return String.format(
+                "It is only possible to configure exactly one of "
+                        + "with%sSerializationSchema, with%<sSerializer or withConfigurable%<sSerializer.",
+                serializerType);
+    }
+
+    private static String exceedsValueSerializerConfigurations() {
+        return exceedsSerializerConfigurations("Value");
+    }
+
+    private static String exceedsKeySerializerConfigurations() {
+        return exceedsSerializerConfigurations("Key");
+    }
```

Comment:
Inline and use `checkState(condition, format, params)`? At the very least, I'd inline the two lower methods.

Reply:
The formatting of `checkState` is very limited and does not use the usual String formatting code. So I would need to replicate the placeholder because `%<s` is not supported. I'd rather stick with the current implementation.

Reply:
Oh I didn't know that. Wtf. Okay leave as is.

Reply:
But please inline the `exceedsValueSerializerConfigurations` and `exceedsKeySerializerConfigurations`; it's really adding not much but noise.

Reply:
One more thing: I would be too lazy to spell out things explicitly. I'd just use `checkState(bla == null, "Key serializer already set")`. That would certainly also make it easier to maintain but of course it requires users that can infer that `withKeyX` are all setting the key serializer ;)